File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    int[][] directions = {{1,0},{-1,0},{0,1},{0,-1}};
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        if(heights == null || heights.length == 0){
            return new ArrayList<>();
        }

        int m = heights.length;
        int n = heights[0].length;

        boolean[][] pacific = new boolean[m][n];
        boolean[][] atlantic = new boolean[m][n];

        List<List<Integer>> result = new ArrayList<>();

        for(int i = 0; i < m; i++){
            dfs(heights,pacific,i,0);
            dfs(heights,atlantic,i,n-1);
        }

        for(int i = 0; i < n; i++){
            dfs(heights,pacific,0,i);
            dfs(heights,atlantic,m-1,i);
        }

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(pacific[i][j] && atlantic[i][j]){
                    result.add(Arrays.asList(i,j));
                }
            }
        }

        return result;
    }
    private void dfs(int[][] grid,boolean[][] ocean, int row, int col){
        ocean[row][col] = true;
        
        for(int[] dir: directions){
            int newR = dir[0] + row;
            int newC = dir[1] + col;
            if(newR >= 0 && newR < grid.length && newC >= 0 && newC < grid[0].length &&
            !ocean[newR][newC] && grid[newR][newC] >= grid[row][col]){
                dfs(grid,ocean,newR,newC);
            }
        }
    }
}

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution(object):
    def pacificAtlantic(self, heights):
        """
        :type heights: List[List[int]]
        :rtype: List[List[int]]
        """
        if not heights or not heights[0]:
            return []

        m, n = len(heights), len(heights[0])

        pacific = [[False] * n for _ in range(m)]
        atlantic = [[False] * n for _ in range(m)]

        result = []
        
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def dfs(grid, ocean, row, col):
            ocean[row][col] = True
            for dir in directions:
                newR, newC = row + dir[0], col + dir[1]
                if 0 <= newR < m and 0 <= newC < n and not ocean[newR][newC] and grid[newR][newC] >= grid[row][col]:
                    dfs(grid, ocean, newR, newC)

        for i in range(m):
            dfs(heights, pacific, i, 0)
            dfs(heights, atlantic, i, n - 1)

        for i in range(n):
            dfs(heights, pacific, 0, i)
            dfs(heights, atlantic, m - 1, i)

        for i in range(m):
            for j in range(n):
                if pacific[i][j] and atlantic[i][j]:
                    result.append([i, j])

        return result
